stan/math/prim/mat/meta/append_return_type.hpp

```diff
@@ -0,0 +1,74 @@
+#ifndef STAN_MATH_PRIM_MAT_META_APPEND_RETURN_TYPE_HPP
+#define STAN_MATH_PRIM_MAT_META_APPEND_RETURN_TYPE_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <vector>
+
+namespace stan {
+  namespace math {
+    /**
+     * This template metaprogram defines the return type for appending
+     * containers of the specified template parameter types.
+     *
+     * This base implementation assumes the template parameters are not
+     * container types and calculates their return type using
+     * <code>return_type</code>.
+     *
+     * Specializations define behavior for other basic types and for containers.
+     *
+     * @tparam T1 First type to be promoted
+     * @tparam T2 Second type to be promoted
+     */
+    template <typename T1, typename T2>
+    struct append_return_type {
+      typedef typename return_type<T1, T2>::type type;
+    };
+
+    /**
+     * If both types are ints, the type member is an int (this is different
+     * than the default Boost type promotion logic used in
+     * <code>return_type</code>).
+     *
+     * @tparam T1 First type to be promoted
+     * @tparam T2 Second type to be promoted
+     */
+    template <>
+    struct append_return_type<int, int> {
+      typedef int type;
+    };
+
+    /**
+     * If both types are Eigen::Matrices with the same Row/Column specification,
+     * then the type member is another Eigen::Matrix with the same Row/Column
+     * specification and a scalar type promoted using <code>return_type</code>.
+     * Part of return type promotion logic for append_array
+     *
+     * @tparam T1 Scalar type of first matrix type
+     * @tparam T2 Scalar type of first matrix type
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     */
+    template <typename T1, typename T2, int R, int C>
+    struct append_return_type<Eigen::Matrix<T1, R, C>,
+                              Eigen::Matrix<T2, R, C> > {
+      typedef typename
+      Eigen::Matrix<typename append_return_type<T1, T2>::type, R, C> type;
```

Comment:
Arguably what you have is more general, but from Stan's perspective, this can just be a regular `return_type` function call because we only allow `double` and autodiff scalars in matrices (not integers).